### ISSUE TITLE
Wheezy to Jessie upgrade left server unable to boot

### ISSUE THREAD

**Issue by macmladen** [omega8cc/boa#972]
But no error was reported during the upgrade in logs.

Even in `/var/log/syslog` states clean `reboot` but it never managed to get online.

Trying to fix now with rescue system.

.barracuda.cnf
https://gist.github.com/macmladen/aa0ed3eb0b4dfdbdb86b8cab9554513f

boa-info
https://gist.github.com/macmladen/5c6d3fbe0a27d5a0035f3095d8a7f8b6

**Comment by omega8cc** [comment]
Where this happened? I mean, where is this VPS hosted?

**Comment by macmladen** [comment]
Hetzner, it is bare metal machine, i7 / 32GB.

Also it is using software RAID for redundancy.

**Comment by omega8cc** [comment]
Hmm.. we have tested this on Linode, DigitalOcean and our VServer instances, and there were no issues. But we have not tested this on bare metal. Although it is weird... Have you tried Squeeze to Wheezy on Hetzner before?

**Comment by omega8cc** [comment]
We should make it clear that we don't recommend to use bare metal, unless you will install your own VPS with VServer on it. Because we don't really test bare metal for a long time.

**Comment by macmladen** [comment]
I mounted **rescue** system, used `chroot` and now trying to find out what I need to do to get it booted.

Everything went as expected (using `screen` and other terminal to monitor progress) and when I gave `reboot` it didn't get online.

Also there is nothing in logs since reboot because it never rebooted.

**No**, I was on Wheezy from the start of this in 2013, before I was on Digital Ocean but never performed major upgrade.

Didn't knew about that on bare metal, but so far everything was as expected. Beside, it should be the same, right?

Anyway, I'd welcome any good idea and continue to work to get it online.

**Comment by macmladen** [comment]
Would some info from some log help?

`cat /etc/debian_version` is **8.5**

**Comment by omega8cc** [comment]
I hope you have a remote console access, and it should provide some hints / display errors when you reboot it via console. We don't know Hetzner configuration for bare metal, so it is rather hard to guess, what could go wrong, and why. Maybe we should get one for crash-testing..

If you have followed the docs, and there were no packages left as not installed, and there were no errors reported, then it should just work.

Or maybe either the hardware is not compatible with Jessie kernel, or the software RAID caused problems.

**Comment by macmladen** [comment]
They are offering Debian 8.4 minimal and LAMP so I guess it should be OK.

I did follow exactly. Not sure about rest but it apparently does not boot at all (grub issue?).

I have tried to use old kernel, `3.2.0-4` instead of `3.16.0-4` but that did not help.

**Comment by omega8cc** [comment]
Now it is probably a good idea to ask your host for assistance, if this is grub somehow broken.

**Comment by macmladen** [comment]
Honestly, I freaked out.

So definitely, it is about RAID and `grub`.

Not sure what was changed upon upgrade but I am going to document it completely here, should  anyone face similar problems.

Server is standalone server and was installer as minimal Wheezy install, current at the time. The disk configuration is 2x3TB. It is configured as RAID1 with swap partition and the rest is used for data.

So two devices
```
/dev/sda
/dev/sdb
```
are joined as redundant on RAID 1 so if one drive fails it is enough just to change it physically and initiate a repair.
```
/dev/md0
/dev/md1
```
are **swap** and **data** partition.

All BOA upgrades went fine for over a year and I was waiting to make a major upgrade to Jessie with 3.1.0 milestone.

Followed steps as advised in [ 26 May 2016 BOA-3.1.0 Full Edition](https://omega8.cc/boa-310-full-edition-380) using `screen` to avoid shell freezing due to inactivity and that went fine in each step up to and including **fifth**.

Then it didn't started at all and was inaccessible by SSH so it is very important that your provider can help you with such situation.

There are two solutions, one is `console` that shows you the boot process so you can repair it from `grub rescue>` terminal or with **boot rescue** system that enables you to boot some other rescue system, mount your RAID and `chroot` into it and then fix.

This is what eventually worked for me: (after booting into rescue system)

```
root@rescue ~ # mount /dev/md1 /mnt
root@rescue ~ # chroot-prepare /mnt ; chroot /mnt
grub-install --target=i386-pc --recheck --debug /dev/sda
update-grub2
```

- first command mounts data partition to mount point `/mnt`
- second prepares and performs `chroot`(preparation is needed to have access to `/dev`)
- third rebuilds `grub` modules and sets boot to `/dev/sda` which I do not think is proper but that finally got system started and I didn't want to risk with experiments but I should probably first tried `/dev/md1`
- fourth finds kernels and register them in device map and writes `/boot/grub/grub.cnf`

The whole business is very tricky and hard to reproduce but when that happens, it is really very bad.

So if anyone needs help, feel free to ask.

**Conclusion:** BOA works perfect on standalone server (bare metal) but some details can make your life more complicated. VPS usually does not cope with **RAID** as infrastructure takes care of that and even that should not be the problem until this major upgrade came.

**Comment by omega8cc** [comment]
Thank you for the update!

We have disabled support for running BOA outside of supported/known virtual machines, to prevent this type of problems in the future: https://github.com/omega8cc/boa/commit/7c4c4a804be8a44037e363108917dbfd739a3658

**Comment by pricejn2** [comment]
@macmladen thanks for posting. I also have a number of bare metal servers that have ran BOA fine for years. Was about to run the Jessie upgrade and saw your ticket. I'll be sure to test the Jessie upgrade path separately on their hardware / Linux builds, seeing the problems you experienced.

**Comment by macmladen** [comment]
Truth to be told, **any major upgrade** is tricky and even on my machine which is powerful i7, with enough memory, very very low load still it took a lot of time as there is a lot of things to be done (I watched closely the river of messages on screen all the time).

As much as BOA build itself and therefore presents a controlled environment there are some things outside of it like a `grub`bootloader and probably some other specifics that someone performs during **minimal install**, some particular image from hosting provider or some custom modification or addition needed to suite some specific needs.

In my case it was just a `grub` as I decided that redundancy is key part of operational server and the moment I experienced one drive failure that enabled me to continue flying with one engine and replace the other in flight (I choose simple partition scheme `swap` and `data` and without LVM as I didn't see the need for such use/complication in my use case scenario).

So I can say it is pretty much safe to be used on standalone server because if you cannot cope with `grub` issues you may not be ready to fly on you own e.g. run unmanaged server. 

I was on shared hosting with regular Drupal installations (like many) then decided to challenge myself first with experimental VPS on Digital Ocean for about a year (I took two months just building server with various options and then few months running just some experimental sites and only then starting very slow migration to BOA and finally giving up on shared hosting) and then also very slowly transitioning from that VPS to standalone, literally like going from fast car into turbo propeller plane (Spitfire) and then taking F-15 (probably the best man made iron bird).

```
BOA [11:17:47] ==> WARN: Wheezy to Jessie upgrade will start in 60 seconds...
BOA [11:17:47] ==> WARN: Now pray it will work... or hit ctrl-c to stop now!
```

I took this as a humorous part (like the `CARD` joke) but I knew something can always turn bad, being BOA or just plain `apt-get update && apt-get upgrade``. As God knew I am atheist I could not pray and it didn't work. But faith also didn't help but knowledge did.

So I would not mark unknown VPS system as `FATAL ERROR` as that sound a bit harsh (whoever is on this plane sure know how to handle troubles), maybe more like a remark that **WARNING: BOA was not architected to be ran on your setup, proceed anyway? (yes/no)** with possible option of a command line switch (that will generate `.allow.any.virt.cnf` anyway although I see no problem with manual file setting except for that **FATAL ERROR**, that sounds like something had really happened while it is just a warning to proceed with caution and to know what you are doing which is by default in our role of system admin.

And lastly, it is just a `grub` and **RAID** that produced a problem for me and I knew it was just that as system didn't boot so it was just a matter of getting a console and/or **rescue system** and do a few commands (there is also `grub rescue terminal` procedure which is tough environment but there is plenty info out there how to diagnose and fix that way).

If I would have enough knowledge I would replicate my situation (standalone server with two drives in soft raid using `mdadm` initially with Wheezy) and then perform **wheezy to jessie** to see what went wrong but as I had drive failure and I **did not** run `update-grub` (I should because it is a new drive in `device-map` and new  `uuid`) maybe that triggered the problem with OS upgrade.

> There is **never** a randomness, only cause and consequence

**Comment by omega8cc** [comment]
We may need to improve the warnings, but we should discourage using bare metal. if you are using bare metal, do yourself a favor, and install VServer. We could even provide a tool to automate this.

**Comment by macmladen** [comment]
What do you think about Docker?

Maybe Dockerized BOA image would be the best way to run on bare-metal?

**Comment by omega8cc** [comment]
No, Docker is not a solution to this problem. Only virtualization like via Linux VServer allows you to safely break your BOA system, easily create local BOA image, re-install it, if needed, etc, all without touching anything in your bare metal, main system.

You could think about running Docker within VPS, but it only adds complexity without any real benefit.

**Comment by macmladen** [comment]
Please do not understand me wrong, I'm trying to learn from this conversation :)

As far as I know **Linux VServer** [[Wikipedia](https://en.wikipedia.org/wiki/Linux-VServer)] is also a kind of **jail mechanism** [[Wikipedia](https://en.wikipedia.org/wiki/Operating-system-level_virtualization)] like Docker which can be seen on table **Implementations**.

Among disadvantages (which counts for both but also for LXC, OpenVZ, Virtuozzo) the main is:

>No clustering or process migration capability is included, so the host kernel and host computer is still a single point of failure for all virtual servers.

As far as I know, Docker was born as improved LXC with the idea of isolating (for instance) services (or clients) under bare-metal so you can have one machine with multiple IPs which runs different services in different containers, one for BOA, another for Mail server, third for Node services, fourth for Ruby apps like Redmine, etc.

Or am I wrong?

**Comment by omega8cc** [comment]
Yes, it is true that Linux VServer is not like some other systems which offer so called 'full virtualization', so fully virtualized VPS can run its own kernel, which comes at the cost of some overhead.

In fact, VServer creates 'containers' (with zero overhead) and not really 'virtual machines'. Its containers are not exactly what Docker does, of course, but it's a similar concept, just much older. We are using it for more years than I dare to admit.

But VServer offers strong enough separation between its 'containers', so you can safely break and destroy them, run major OS upgrades within their contexts, etc, and you will not affect the host system.

This is not possible with Docker (which is expected to run on bare metal to make sense), so VServer offers both very lightweight virtualization (much easier to manage than Docker with its myriad of tools, by the way), and separation which makes running different systems like completely separate environments.

**Comment by attiks** [comment]
Since you mention Hetzner and bare metal, I'm using them for years using boa, I once had a similar problem and it was unrelated to boa, nowadays I'm mostly using hardware raid, but I ran it at least on 10-15 software RAIDs in the past.

**Comment by macmladen** [comment]
So if you can throw some preparatory script that prepares bare-metal and puts up VServer to host BOA that would be truly awesome.

Also I'd be grateful if you can provide some practical tutorial that could help me in the process as I just **adore** BOA but also may need Redmine, Gitlab, Mail all of which may run in their own containers (VServers).

I was looking for some solution to have Redmine and/or mail in separate containers so I wanted LXC as it seemed lightweight but then I was wondering if Docker is just toolset for LXC or does it offer more than just that.

**Comment by attiks** [comment]
Count me in, would love to give it a try

**Comment by omega8cc** [comment]
OK, please open feature request, then :)

It's a good idea to open source this tiny and simple tool we use to set up VServer containers. We just need to make it hardware agnostic, so it could work on any bare metal. You could then use `xboa` to migrate your Octopus instances away from host to guest OS.

**Comment by EdNett** [comment]
Hello, Would you please provide some clarification?:
I thought boa was supposed to be built on a "dedicated box" (or a vps), with a minimal install of Debian Jessie. Is this also called "bare metal"? Or is bare metal something less than that - just an ssh server/client for example?

If you are ststing that a stand-alone dedicated server with jessie installed on it is not sufficient to install and run boa, please state so very clearly - because most of us believe and have believed that this was the recommended way to install and use boa for years. And that boa could be the only thing on the vps or dedicated box (with the exception of a few hand coded or wordpress sites, etc) - i.e. that we should not try to run a mail server or other serious applications on a boa server.

Is this all changing now?

If it is, I would like to see docker used, and will work towards coming up with an implementation of it (I guess the networking / IP will be the problem) UNLESS you are really set against it. I don't want to become dependent on docker, per se, but as long as it is just a toolset built around clusterable lxc containers and remains open source, it seems to be the best implementation. I will look at v-server, as I await your clarification.  Thanks for boa and your great work. I am just requesting clarification for all that Jessie installed directly on a dedicated server is no longer the suggested way of installing and running boa. Thanks.

**Comment by omega8cc** [comment]
@EdNett Using bare metal (running BOA directly on a dedicated box with no virtualization layer) was never recommended, just listed as "supported", although we have never tested it and we have never used it. It is just an awfully bad idea.

Now it is clearly marked as unsupported when you will try to run BOA on it (it is already implemented in HEAD).

Bare metal (running BOA directly on a dedicated box with no virtualization layer) can add unexpected complexity due to things like hardware components, software RAID etc, and may brick your system on major OS upgrade, for reason we couldn't predict nor expect -- see this very issue.

We don't have any immediate plans to support Docker, not because of preference or like/dislike. Docker makes sense only on bare metal (otherwise you gain no benefit, only extra complexity, plus double virtualization overhead), so you are back in the same corner, and you could break your system on major OS upgrade, very easily.

If you want to use dedicated box, use VServer containers.

**Comment by omega8cc** [comment]
I should add that running virtualization on the service level (Docker container) in contrast to OS virtualization (any VPS) would require changes to Aegir itself, because currently Aegir assumes "monolithic" OS with services, and not services running within Docker containers.

**Comment by bserem** [comment]
Follow up of #988 I'd like to update people that with the release of BOA v3.1.1 you need to select
`Paravirtualization interface: Legacy`
in Virtualbox to be able to install BOA.

**Comment by omega8cc** [comment]
What is the response on Virtualbox without that option if you run `virt-what` command after installing BOA wrappers?

**Comment by bserem** [comment]
Here it is:

```
root@debianvm:~# virt-what
virtualbox
kvm
root@debianvm:~# boa in-stable local my@email none php-5.6

=== OOPS! ===

You are running not supported virtualization system:
   virtualbox kvm

If you wish to try BOA on this system anyway,
please create an empty control file:
   /root/.allow.any.virt.cnf

Please be aware that it may not work at all,
or you can experience errors breaking BOA.

BOA is known to work well on:
  * Linux KVM guest
  * Linux VServer guest
  * Parallels guest
  * VirtualBox guest
  * VMware ESXi guest
  * Xen guest

BOA is not designed to run directly on a bare metal.
We recommend to install Linux VServer if you plan to
use BOA on a dedicated machine -- server or laptop.

Note: There are plans to support Linux VServer installation
via BOA tools: https://github.com/omega8cc/boa/issues/980

Bye

root@debianvm:~#
```

**Comment by bserem** [comment]
Hmmm... comment above didn't get a 'code' block, but it is readable I believe.

Btw, with Legacy mode in Virtualbox I get this answer:
```
root@debianvm:~# virt-what
virtualbox
```

**Comment by omega8cc** [comment]
Thanks. Maybe installed `virt-what` gets confused, is outdated, or just not ready to properly detect this environment. We could add a workaround, like we did for xen, I think.

**Comment by macmladen** [comment]
Just one more addition which may not be obvious: problem arouse due to **software RAID** otherwise everything was working on bare metal as would normally in any kind of VPS. I suppose that **hardware** (or proper) RAID would not face such problems but there are too many options to be sure. 

However, both technologies are important as they make your data safe. This is setting that makes sense only on bare-metal (standalone) server.

**Comment by omega8cc** [comment]
Not sure if this is related, but who knows: https://github.com/omega8cc/boa/issues/912#issuecomment-231088287
